In [37]:
import pandas as pd
from scipy.stats import shapiro, fligner, ttest_ind

In [ ]:
# Recursos:
# Comparación de media (t-test)
# https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html
# https://towardsdatascience.com/inferential-statistics-series-t-test-using-numpy-2718f8f9bf2f

# Comprobación de normalidad
# https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.shapiro.html

# Comparación de homocedasticidad
# https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.fligner.html
# https://docs.scipy.org/doc/scipy-0.19.1/reference/generated/scipy.stats.fligner.html

# Comparaciones de distr no parametricos
# https://machinelearningmastery.com/nonparametric-statistical-significance-tests-in-python/
# https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.wilcoxon.html
# https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.mannwhitneyu.html

In [2]:
df = pd.read_csv('train.csv')

In [3]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

# Test de normalidad

Para comprobar la normalidad de las variables numéricas Age y Fare aplicaremos el test Shapiro-Wilk. La hipótesis nula es que la variable sigue una distribución normal. El nivel de significancia será $\alpha = 0.05$, si el p-valor es inferior a este, rechazaremos la hipótesis nula.

In [18]:
# test Saphiro Wilk para comprobar la normalidad de Age
print("Nº de observaciones:",  len(df['Age'].dropna()))
shapiro(df['Age'].dropna())

Nº de observaciones: 714


(0.9814548492431641, 7.322165629375377e-08)

In [19]:
# test Saphiro Wilk para comprobar la normalidad de Fare
print("Nº de observaciones:",  len(df['Fare'].dropna()))
shapiro(df['Fare'].dropna())

Nº de observaciones: 891


(0.5218914747238159, 1.0789998175301091e-43)

Ninguna de las dos variables estudiadas sigue una distribución normal. Sin embargo, dado que tenemos un gran número de observaciones (N = (714, 891) > 20) podemos asumir el teorema central del límite. La distribución de la media de las variables estudidas, gracias a la gran cantidad de observaciones, seguirá una distribución normal con una media de población $\mu$ y una varianza $\frac{\sigma^2}{N}$.

#  Test de homocedasticidad

Para comprobar la homocedasticidad de las variables usaremos el test de Fligner-Killeen, dado que nuestros datos no siguen una distribución normal. La hipótesis nula asume igualdad de varianza.

In [22]:
sobrevive = df['Survived'] == 1
nop = df['Survived'] == 0

hombre = df['Sex'] == 'male'
mujer = df['Sex'] == 'female'

niño = df['Age'] <= 12
adolescente = (df['Age'] > 12 & df['Age'] < 18)
adulto = df['Age'] >= 18

primera = df['Pclass'] == 1
segunda = df['Pclass'] == 2
tercera = df['Pclass'] == 3

In [51]:
edadvssupervivencia = [df.loc[sobrevive, 'Age'].dropna() , df.loc[nop, 'Age'].dropna()]
edadvsclase = [df.loc[primera, 'Age'].dropna() , df.loc[segunda, 'Age'].dropna(), df.loc[tercera, 'Age'].dropna()]

In [52]:
billetevssupervivencia = [df.loc[sobrevive, 'Fare'].dropna() , df.loc[nop, 'Fare'].dropna()]
billetevsclase = [df.loc[primera, 'Fare'].dropna() , df.loc[segunda, 'Fare'].dropna(), df.loc[tercera, 'Fare'].dropna()]

In [53]:
fligner(*edadvssupervivencia)

FlignerResult(statistic=1.0436060696132083, pvalue=0.3069842761053375)

In [54]:
fligner(*edadvsclase)

FlignerResult(statistic=11.317210079614878, pvalue=0.0034873782605697604)

In [55]:
fligner(*billetevssupervivencia)

FlignerResult(statistic=96.2527395111909, pvalue=1.0111515793725321e-22)

In [56]:
fligner(*billetevsclase)

FlignerResult(statistic=365.7983079876698, pvalue=3.697487170015036e-80)

Todos los casos tienen pvalues > 0.05 por lo que podemos asumir homocedasticidad para todos los casos.

# Análisis estadístico

##  Comparación de medias (t-test)

$H_0: \mu_1 = \mu_2 $

In [57]:
# Edad supervivientes vs no
ttest_ind(*edadvssupervivencia)

Ttest_indResult(statistic=-2.06668694625381, pvalue=0.03912465401348249)

In [58]:
# Edad 1era clase vs 2da clase
ttest_ind(edadvsclase[0], edadvsclase[1])

Ttest_indResult(statistic=5.485187676773201, pvalue=7.835568991415144e-08)

In [59]:
# Edad 2da clase vs 3ra clase
ttest_ind(edadvsclase[1], edadvsclase[2])

Ttest_indResult(statistic=3.927800191020872, pvalue=9.715078600777852e-05)

In [60]:
# Edad 1era clase vs 3ra clase
ttest_ind(edadvsclase[0], edadvsclase[2])

Ttest_indResult(statistic=10.849122601201033, pvalue=6.134470007830625e-25)

In [61]:
# Billete supervivientes vs no
ttest_ind(*billetevssupervivencia)

Ttest_indResult(statistic=7.939191660871055, pvalue=6.120189341924198e-15)

In [62]:
# Billete 1ra clase vs 2da clase
ttest_ind(billetevsclase[0], billetevsclase[1])

Ttest_indResult(statistic=10.851594207309702, pvalue=3.303644173119574e-24)

In [63]:
# Billete 2da clase vs 3ra clase
ttest_ind(billetevsclase[1], billetevsclase[2])

Ttest_indResult(statistic=6.600620525446866, pvalue=8.304239675888693e-11)

In [64]:
# Billete 1ra clase vs 3ra clase
ttest_ind(billetevsclase[0], billetevsclase[2])

Ttest_indResult(statistic=19.448650285047485, pvalue=8.908959353672071e-68)

##  Wilcox o Mann-Whitney

In [65]:
# revisar apuntes estadística avanzada ->¿test previo para decidir uno u otro? ¿Eran datos apareados?

#  Camparación de más de dos grupos (nuestro caso) - ANOVA

In [ ]:
# WIP